### PR TITLE
Chapter 11

### PR DESCRIPTION
**Pull request by Moonforeva**

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: Gemfile.lock [hunk 1]
```diff
@@ -270,6 +272,7 @@ GEM
     websocket-driver (0.7.3)
       websocket-extensions (>= 0.1.0)
     websocket-extensions (0.1.5)
+    will_paginate (3.3.0)
```
Comment (saobv-2223) [inline Gemfile.lock:275]: using gem kaminari instead of will_paginate to paginate

File: app/assets/stylesheets/account_activations.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/account_activations.scss:3]: remove file that never used

File: app/controllers/account_activations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class AccountActivationsController < ApplicationController
+
+  def edit
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb:3]:
```suggestion
class AccountActivationsController < ApplicationController
  def edit
```

File: app/controllers/account_activations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    user = User.find_by(email: params[:email])
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb:4]: remove ()

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,11 +1,20 @@
 class UsersController < ApplicationController
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:2]: resolve conflict, please

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,11 +1,20 @@
...
   def index
     @users = User.page(params[:page]).per Settings.per_page
+=======
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :admin_user, only: :destroy
+
+  def index
+    @users = User.paginate(page: params[:page])
+>>>>>>> 02480dc... Chapter 10
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:17]: resolve conflict

File: app/mailers/user_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation(user)
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb:3]:
```suggestion
class UserMailer < ApplicationMailer
  def account_activation user
```

File: app/models/user.rb [hunk 8]
```diff
@@ -33,7 +34,8 @@ def remember
     update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated?(remember_token)
+  def authenticated?(attribute, token))
+    digest = send("#{attribute}_digest")
```
Comment (saobv-2223) [inline app/models/user.rb:38]: remove ()

File: app/models/user.rb [hunk 9]
```diff
@@ -43,9 +45,23 @@ def forgets
     update_attribute :remember_digest, nil
   end
 
+  def activate
+    update_attribute(:activated, true)
+    update_attribute(:activated_at, Time.zone.now)
+  end
```
Comment (saobv-2223) [inline app/models/user.rb:51]: remove ()

File: app/views/user_mailer/account_activation.html.erb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<h1>Sample App</h1>
+<p>Hi <%= @user.name %>,</p>
+<p>
+Welcome to the Sample App! Click on the link below to activate your account:
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.html.erb:4]:
```suggestion
  Welcome to the Sample App! Click on the link below to activate your account:
```

File: app/views/user_mailer/account_activation.html.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+</p>
+<%= link_to "Activate", edit_account_activation_url(@user.activation_token, email: @user.email) %>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.html.erb:6]:
```suggestion
<%= link_to "Activate", edit_account_activation_url(@user.activation_token, email: @user.email) %>

```

File: app/views/user_mailer/account_activation.text.erb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Hi <%= @user.name %>,
+Welcome to the Sample App! Click on the link below to activate your account:
+<%= edit_account_activation_url(@user.activation_token, email: @user.email) %>
```
Comment (saobv-2223) [inline app/views/user_mailer/account_activation.text.erb:3]:
```suggestion
<%= edit_account_activation_url(@user.activation_token, email: @user.email) %>

```

File: app/views/users/edit.html.erb [hunk 13]
```diff
@@ -21,7 +21,11 @@
     <% end %>
     <div class="gravatar_edit">
       <%= gravatar_for @user %>
+<<<<<<< HEAD
```
Comment (saobv-2223) [inline app/views/users/edit.html.erb:24]: resolve conflict

File: config/routes.rb [hunk 14]
```diff
@@ -10,4 +10,5 @@
   post "/login", to: "sessions#create"
   delete "/logout", to: "sessions#destroy"
   resources :users
+  resources :account_activations, only: [:edit]
```
Comment (saobv-2223) [inline config/routes.rb:13]:
```suggestion
  resources :account_activations, only: :edit
```

**Comment by Moonforeva** [comment]
Ready

File: app/mailers/application_mailer.rb [hunk 15]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: "from@example.com"
+  default from: "noreply@example.com"
```
Comment (saobv-2223) [inline app/mailers/application_mailer.rb:2]: using gem figaro to manage enviroment variables

File: app/models/user.rb [hunk 16]
```diff
@@ -43,9 +45,23 @@ def forgets
     update_attribute :remember_digest, nil
   end
 
+  def activate
+    update_attribute :activated, true
+    update_attribute :activated_at, Time.zone.now
```
Comment (saobv-2223) [inline app/models/user.rb:50]:
```suggestion
    update activated: true, activated_at: Time.zone.now
```

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/sessions_controller.rb [hunk 17]
```diff
@@ -4,9 +4,16 @@ def new; end
...
-      params [:session][:remember_me] == Settings.remember_me ? remember(user) : forget(user)
-      redirect_to user
+      if user.activated?
+        log_in user
+        params [:session][:remember_me] == Settings.remember_me ? remember(user) : forget(user)
+        redirect_back_or user
+      else
+        message = "Account not activated. "
+        message += "Check your email for the activation link."
+        flash[:warning] = message
```
Comment (vanvtt-0952) [inline app/controllers/sessions_controller.rb:14]: `message = "Account not activated. "`
` message += "Check your email for the activation link."`
`flash[:warning] = message`
change to
` flash[:warning] = Account not activated. Check your email for the activation link.`

File: app/helpers/account_activations_helper.rb [hunk 18]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (vanvtt-0952) [inline app/helpers/account_activations_helper.rb:1]: delete this file plz

File: config/environments/development.rb [hunk 19]
```diff
@@ -31,9 +31,10 @@
   # Store uploaded files on the local file system (see config/storage.yml for options).
   config.active_storage.service = :local
 
-  # Don't care if the mailer can't send.
+  # Don"t care if the mailer can"t send.
   config.action_mailer.raise_delivery_errors = false
-
+  host = "localhost:3000"
```
Comment (vanvtt-0952) [inline config/environments/development.rb:36]: config this project to use send a mail with Gmail plz,

image

config.action_mailer.raise_delivery_errors = true
  config.action_mailer.delivery_method = :smtp
  host = "localhost"
  config.action_mailer.default_url_options = { host: host, port: '3000', protocol: 'http' }
  ActionMailer::Base.smtp_settings = {
    address: 'smtp.gmail.com',
    port: '587',
    authentication: :plain,
    user_name: ENV[ "YOUR_USERNAME" ],
    password: ENV[ "YOUR_PASSWORD" ],
    enable_starttls_auto:  true
  }

**Comment by Moonforeva** [comment]
Ready

**Comment by manhbnt-0960** [comment]
merged

**Pull request merged**